controllers/mysql_combined_test.go

```diff
@@ -99,3 +109,70 @@ func TestMySQLHappyPath(t *testing.T) {
 	EnsureDelete(ctx, t, tc, mySQLServerInstance)
 	EnsureDelete(ctx, t, tc, mySQLReplicaInstance)
 }
+
+func RunMySQLUserHappyPath(ctx context.Context, t *testing.T, mySQLServerName string, mySQLDBName string, rgName string) {
+	assert := assert.New(t)
+
+	// Create a user in the DB
+	username := GenerateTestResourceNameWithRandom("user", 10)
+	user := &azurev1alpha1.MySQLUser{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      username,
+			Namespace: "default",
+		},
+		Spec: azurev1alpha1.MySQLUserSpec{
+			ResourceGroup: rgName,
+			Server:        mySQLServerName,
+			DbName:        mySQLDBName,
+			Username:      username,
+			Roles:         []string{"SELECT"},
+		},
+	}
+	EnsureInstance(ctx, t, tc, user)
+
+	// Update user
+	namespacedName := types.NamespacedName{Name: username, Namespace: "default"}
+	err := tc.k8sClient.Get(ctx, namespacedName, user)
+	assert.NoError(err)
+
+	updatedRoles := []string{"UPDATE", "DELETE", "CREATE", "DROP"}
+	user.Spec.Roles = updatedRoles
+	err = tc.k8sClient.Update(ctx, user)
+	assert.NoError(err)
+
+	// TODO: Ugh this is fragile, the path to the secret should probably be set on the status?
+	secretNamespacedName := types.NamespacedName{Name: mySQLServerName, Namespace: "default"}
+	adminSecret, err := tc.secretClient.Get(ctx, secretNamespacedName)
+	adminUser := string(adminSecret["fullyQualifiedUsername"])
+	adminPassword := string(adminSecret[mysqluser.MSecretPasswordKey])
+	fullServerName := string(adminSecret["fullyQualifiedServerName"])
+
+	db, err := mysql.ConnectToSqlDB(
+		ctx,
+		mysql.MySQLDriverName,
+		fullServerName,
+		mySQLDBName,
+		mysql.MySQLServerPort,
+		adminUser,
+		adminPassword)
+	assert.NoError(err)
+
+	assert.Eventually(func() bool {
+		roles, err := mysql.ExtractUserRoles(ctx, username, db)
+		assert.NoError(err)
+
+		rolesForThisDB := roles[mySQLDBName]
+
+		if len(rolesForThisDB) != len(updatedRoles) {
+			return false
+		}
+
+		for _, role := range updatedRoles {
+			if _, ok := rolesForThisDB[role]; !ok {
+				return false
+			}
+		}
+
+		return true
+	}, tc.timeout, tc.retry, "waiting for DB user to be updated")
```

Comment:
Test for removing a user's role as well?

Reply:
This test actually tests both at once. I set the desired roles from `SELECT` to `UPDATE, DELETE, CREATE, DROP`

Reply:
So for the test to pass `SELECT` had to have been removed.

Reply:
Oh sorry, missed that!